Lib/glyphsLib/classes.py

```diff
@@ -3727,3 +3806,77 @@ def removeKerningForPair(self, fontMasterId, leftKey, rightKey, direction=LTR):
             del self._kerning[fontMasterId][leftKey]
         if not self._kerning[fontMasterId]:
             del self._kerning[fontMasterId]
+
+
+Number = Union[int, float]
+
+
+class LayerPointPen(AbstractPointPen):
+    """A point pen to draw onto GSLayer object.
+
+    See :mod:`fontTools.pens.basePen` and :mod:`fontTools.pens.pointPen` for an
+    introduction to pens.
+    """
+
+    __slots__ = ("_layer", "_path")
+
+    def __init__(self, layer: GSLayer) -> None:
+        self._layer: GSLayer = layer
+        self._path: Optional[GSPath] = None
+
+    def beginPath(self, **kwargs: Any) -> None:
+        if self._path is not None:
+            raise ValueError("Call endPath first.")
+
+        self._path = GSPath()
+        self._path.closed = True  # Until proven otherwise.
+
+    def endPath(self) -> None:
+        if self._path is None:
+            raise ValueError("Call beginPath first.")
+
+        if self._path.closed:
+            self._path.nodes.append(self._path.nodes.pop(0))
+        self._layer.paths.append(self._path)
+        self._path = None
+
+    def addPoint(
+        self,
+        pt: Tuple[Number, Number],
+        segmentType: Optional[str] = None,
+        smooth: bool = False,
+        name: Optional[str] = None,
+        userData: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ) -> None:
+        if self._path is None:
+            raise ValueError("Call beginPath first.")
+
+        if segmentType == "move":
+            if self._path.nodes:
+                raise ValueError("For an open contour, 'move' must come first.")
+            self._path.closed = False
+
+        node = GSNode(
+            Point(*pt),
+            nodetype=_to_glyphs_node_type(segmentType),
+            smooth=smooth,
+            name=name,
+        )
+        if userData:
+            node.userData = userData
+        self._path.nodes.append(node)
+
+    def addComponent(
+        self, baseGlyph: str, transformation: Transform, **kwargs: Any,
```

Comment:
transformation should be a generic 6-tuple in the pointPen protocol, you can turn that to glyphsLib's own Transform class inside the method before initialising GSComponent